### PR TITLE
[RFC;WIP] Catching numerical errors during sampling

### PR DESCRIPTION
**Pull request by yebai**
This PR addresses the issue occurred when running the following model (see https://github.com/TuringLang/Turing.jl/pull/709#issuecomment-471324281).

```julia

# Simulate data
using Distributions

V = 5   # vocab
K = 3   # topic
M = 25  # doc

alpha = collect(ones(K) / K);
beta = collect(ones(V) / V);

theta = rand(Dirichlet(alpha), M)
phi = rand(Dirichlet(beta), K)
@info "true vals" phi theta sum(phi * theta, dims=1)

avg_doc_length = 50

doc_length = rand(Poisson(avg_doc_length), M)

N = sum(doc_length)

w = Vector{Int}(undef, N)
doc = Vector{Int}(undef, N)

n = 1
for m = 1:M, i=1:doc_length[m]
    global n
    z = rand(Categorical(theta[:,m]))
    w[n] = rand(Categorical(phi[:,z]))
    doc[n] = m
    n = n + 1
end

# LDA
using Turing

struct CategoricalReal<: Distributions.DiscreteUnivariateDistribution
    p
end

function Distributions.logpdf(d::CategoricalReal, k::Int64)
    return log(d.p[k])
end

@model lda_collapsed(K, V, M, N, w, doc, beta, alpha) = begin
    theta = Matrix{Real}(undef, K, M)
    for m = 1:M
        theta[:,m] ~ Dirichlet(alpha)
    end

    phi = Matrix{Real}(undef, V, K)
    for k = 1:K
        phi[:,k] ~ Dirichlet(beta)
    end

    word_dist = phi * theta

    for n = 1:N
        w[n] ~ CategoricalReal(word_dist[:,doc[n]])
    end
end

mf = lda_collapsed(K, V, M, N, w, doc, beta, alpha)

# c3= sample(mf, Turing.NUTS(2000, 0.65))
c3= sample(mf, HMCDA(2000, 0.65, 1.0))
```

### CODE REVIEW

**Comment by yebai** [comment]
Example output

```julia
julia> c3= sample(mf, HMCDA(2000, 0.65, 1.0))
[ Info: [Turing] looking for good initial eps...
[ Info: [Turing] found initial ϵ: 0.215625
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error in gradients. Rejecting current proposal...
└ @ Turing.Core ~/.julia/dev/Turing/src/core/ad.jl:169
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error has been detected: DomainError. Rejecting current proposal...
└ @ Turing.Core.VarReplay ~/.julia/dev/Turing/src/core/VarReplay.jl:118
┌ Warning: Numerical error in gradients. Rejecting current proposal...
└ @ Turing.Core ~/.julia/dev/Turing/src/core/ad.jl:169
┌ Warning: Numerical error in gradients. Rejecting current proposal...
└ @ Turing.Core ~/.julia/dev/Turing/src/core/ad.jl:169
[ Info:  Adapted ϵ = 0.21713136065648764, std = [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]; 1000 iterations is used for adaption.
[HMCDA] Sampling...100% Time: 0:07:21
[HMCDA] Finished with
  Running time        = 440.3711882570004;
  Accept rate         = 0.6765;
  #lf / sample        = 5.4355;
  #evals / sample     = 59.0015;
  pre-cond. metric    = [1.0, 1.0, 1.0, 1.0, 1.0, 1.0,....
Object of type Chains, with data of type 2000×96×1 Array{Union{Missing, Float64},3}

Log evidence      = 0.0
Iterations        = 1:2000
Thinning interval = 1
Chains            = 1
Samples per chain = 2000
internals         = elapsed, epsilon, eval_num, lf_eps, lf_num, lp
parameters        = theta[Colon(), 25][3], theta[Colon(), 4][2], theta[Colon(), 16][2], phi[Colon(), 1][2], theta[Colon(), 14][2], phi[Colon(), 3][2], phi[Colon(), 3][3], phi[Colon(), 2][3], phi[Colon(), 3][1], theta[Colon(), 18][2], theta[Colon(), 22][3], phi[Colon(), 3][5], theta[Colon(), 9][3], theta[Colon(), 24][2], theta[Colon(), 15][3], theta[Colon(), 4][1], theta[Colon(), 8][2], theta[Colon(), 16][3], phi[Colon(), 2][5], theta[Colon(), 21][3], theta[Colon(), 18][1], theta[Colon(), 11][3], theta[Colon(), 2][3], theta[Colon(), 16][1], theta[Colon(), 6][3], theta[Colon(), 24][1], theta[Colon(), 3][3], theta[Colon(), 12][2], theta[Colon(), 7][3], theta[Colon(), 5][2], theta[Colon(), 9][1], theta[Colon(), 7][2], theta[Colon(), 8][1], theta[Colon(), 4][3], theta[Colon(), 15][1], theta[Colon(), 9][2], theta[Colon(), 23][2], theta[Colon(), 15][2], theta[Colon(), 5][3], theta[Colon(), 17][1], theta[Colon(), 7][1], phi[Colon(), 2][2], theta[Colon(), 18][3], theta[Colon(), 24][3], phi[Colon(), 1][5], theta[Colon(), 1][1], theta[Colon(), 20][3], theta[Colon(), 3][1], theta[Colon(), 13][2], theta[Colon(), 1][2], theta[Colon(), 14][1], theta[Colon(), 17][3], theta[Colon(), 20][1], theta[Colon(), 11][1], theta[Colon(), 17][2], theta[Colon(), 11][2], theta[Colon(), 6][1], theta[Colon(), 13][1], theta[Colon(), 22][2], theta[Colon(), 25][1], phi[Colon(), 2][4], theta[Colon(), 6][2], theta[Colon(), 23][3], theta[Colon(), 12][3], theta[Colon(), 2][2], theta[Colon(), 21][1], phi[Colon(), 1][3], theta[Colon(), 19][2], theta[Colon(), 3][2], theta[Colon(), 10][3], phi[Colon(), 2][1], theta[Colon(), 12][1], theta[Colon(), 10][1], theta[Colon(), 10][2], theta[Colon(), 21][2], theta[Colon(), 23][1], theta[Colon(), 14][3], theta[Colon(), 2][1], phi[Colon(), 1][1], phi[Colon(), 1][4], phi[Colon(), 3][4], theta[Colon(), 20][2], theta[Colon(), 13][3], theta[Colon(), 25][2], theta[Colon(), 1][3], theta[Colon(), 5][1], theta[Colon(), 19][3], theta[Colon(), 19][1], theta[Colon(), 8][3], theta[Colon(), 22][1]

parameters
                       Mean    SD   Naive SE  MCSE     ESS   
   phi[Colon(), 1][1] 0.0332 0.0080   0.0002 0.0002 1220.9107
   phi[Colon(), 1][2] 0.9256 0.0410   0.0009 0.0050   68.4665
   phi[Colon(), 1][3] 0.0010 0.0046   0.0001 0.0003  321.7259
   phi[Colon(), 1][4] 0.0331 0.0336   0.0008 0.0043   60.1027
   phi[Colon(), 1][5] 0.0070 0.0084   0.0002 0.0008  100.0046
   phi[Colon(), 2][1] 0.0013 0.0029   0.0001 0.0002  185.5639
   phi[Colon(), 2][2] 0.0021 0.0103   0.0002 0.0006  341.2070
   phi[Colon(), 2][3] 0.8577 0.0459   0.0010 0.0047   96.7461
   phi[Colon(), 2][4] 0.0528 0.0377   0.0008 0.0048   62.0012
   phi[Colon(), 2][5] 0.0861 0.0183   0.0004 0.0008  478.8557
   phi[Colon(), 3][1] 0.0016 0.0052   0.0001 0.0003  410.5673
   phi[Colon(), 3][2] 0.0125 0.0174   0.0004 0.0021   71.4694
   phi[Colon(), 3][3] 0.0067 0.0151   0.0003 0.0013  136.5771
   phi[Colon(), 3][4] 0.8563 0.0346   0.0008 0.0028  151.7590
   phi[Colon(), 3][5] 0.1228 0.0210   0.0005 0.0017  153.0888
 theta[Colon(), 1][1] 0.6724 0.0620   0.0014 0.0025  636.9136
 theta[Colon(), 1][2] 0.3099 0.0605   0.0014 0.0027  487.8954
 theta[Colon(), 1][3] 0.0176 0.0232   0.0005 0.0024   95.2822
 theta[Colon(), 2][1] 0.9365 0.0432   0.0010 0.0045   92.7245
 theta[Colon(), 2][2] 0.0287 0.0269   0.0006 0.0017  265.1647
 theta[Colon(), 2][3] 0.0348 0.0374   0.0008 0.0052   51.7033
 theta[Colon(), 3][1] 0.0340 0.0420   0.0009 0.0025  286.5526
 theta[Colon(), 3][2] 0.0375 0.0331   0.0007 0.0030  120.1366
 theta[Colon(), 3][3] 0.9285 0.0560   0.0013 0.0039  202.8153
 theta[Colon(), 4][1] 0.7749 0.0609   0.0014 0.0040  231.2429
 theta[Colon(), 4][2] 0.1817 0.0532   0.0012 0.0020  684.9597
 theta[Colon(), 4][3] 0.0434 0.0421   0.0009 0.0043   94.2651
 theta[Colon(), 5][1] 0.0078 0.0239   0.0005 0.0013  329.2717
 theta[Colon(), 5][2] 0.8583 0.0768   0.0017 0.0074  108.6121
 theta[Colon(), 5][3] 0.1339 0.0735   0.0016 0.0069  112.2860
 theta[Colon(), 6][1] 0.7405 0.0711   0.0016 0.0042  286.2960
 theta[Colon(), 6][2] 0.0722 0.0402   0.0009 0.0015  682.1803
 theta[Colon(), 6][3] 0.1873 0.0676   0.0015 0.0048  198.4854
 theta[Colon(), 7][1] 0.0085 0.0261   0.0006 0.0017  237.7557
 theta[Colon(), 7][2] 0.3328 0.0646   0.0014 0.0032  399.9353
 theta[Colon(), 7][3] 0.6587 0.0685   0.0015 0.0033  420.1104
 theta[Colon(), 8][1] 0.0908 0.0592   0.0013 0.0042  200.9364
 theta[Colon(), 8][2] 0.0866 0.0516   0.0012 0.0024  451.5781
 theta[Colon(), 8][3] 0.8226 0.0791   0.0018 0.0052  235.3652
 theta[Colon(), 9][1] 0.0071 0.0171   0.0004 0.0010  275.2505
 theta[Colon(), 9][2] 0.9800 0.0302   0.0007 0.0022  180.3047
 theta[Colon(), 9][3] 0.0129 0.0240   0.0005 0.0022  122.1070
theta[Colon(), 10][1] 0.1587 0.0540   0.0012 0.0014 1453.8745
theta[Colon(), 10][2] 0.6454 0.0794   0.0018 0.0050  251.3350
theta[Colon(), 10][3] 0.1959 0.0725   0.0016 0.0050  210.1872
theta[Colon(), 11][1] 0.4743 0.0742   0.0017 0.0046  260.4840
theta[Colon(), 11][2] 0.0050 0.0119   0.0003 0.0009  182.6169
theta[Colon(), 11][3] 0.5207 0.0761   0.0017 0.0050  231.8035
theta[Colon(), 12][1] 0.7976 0.0675   0.0015 0.0044  238.2718
theta[Colon(), 12][2] 0.0076 0.0141   0.0003 0.0009  227.9586
theta[Colon(), 12][3] 0.1949 0.0674   0.0015 0.0047  203.1343
theta[Colon(), 13][1] 0.2787 0.0665   0.0015 0.0026  680.2261
theta[Colon(), 13][2] 0.0968 0.0450   0.0010 0.0020  484.3061
theta[Colon(), 13][3] 0.6245 0.0740   0.0017 0.0033  511.1989
theta[Colon(), 14][1] 0.0331 0.0459   0.0010 0.0037  155.0682
theta[Colon(), 14][2] 0.0264 0.0254   0.0006 0.0017  221.9285
theta[Colon(), 14][3] 0.9405 0.0542   0.0012 0.0041  179.0258
theta[Colon(), 15][1] 0.5590 0.0754   0.0017 0.0048  242.3751
theta[Colon(), 15][2] 0.1532 0.0564   0.0013 0.0014 1718.6642
theta[Colon(), 15][3] 0.2879 0.0767   0.0017 0.0052  216.0320
theta[Colon(), 16][1] 0.5046 0.0730   0.0016 0.0031  566.8263
theta[Colon(), 16][2] 0.4333 0.0752   0.0017 0.0045  280.4898
theta[Colon(), 16][3] 0.0621 0.0537   0.0012 0.0042  166.5813
theta[Colon(), 17][1] 0.9832 0.0247   0.0006 0.0013  373.5230
theta[Colon(), 17][2] 0.0066 0.0172   0.0004 0.0010  314.2108
theta[Colon(), 17][3] 0.0102 0.0149   0.0003 0.0015   99.1602
theta[Colon(), 18][1] 0.9379 0.0510   0.0011 0.0060   71.8590
theta[Colon(), 18][2] 0.0077 0.0171   0.0004 0.0011  230.1507
theta[Colon(), 18][3] 0.0543 0.0476   0.0011 0.0062   59.2249
theta[Colon(), 19][1] 0.1031 0.0492   0.0011 0.0016  897.1988
theta[Colon(), 19][2] 0.6671 0.0877   0.0020 0.0055  252.9869
theta[Colon(), 19][3] 0.2298 0.0832   0.0019 0.0051  264.0054
theta[Colon(), 20][1] 0.1740 0.0613   0.0014 0.0019 1008.7599
theta[Colon(), 20][2] 0.6832 0.0864   0.0019 0.0057  231.3089
theta[Colon(), 20][3] 0.1428 0.0719   0.0016 0.0063  129.1709
theta[Colon(), 21][1] 0.4886 0.0623   0.0014 0.0028  496.1107
theta[Colon(), 21][2] 0.4794 0.0656   0.0015 0.0046  199.7942
theta[Colon(), 21][3] 0.0321 0.0354   0.0008 0.0039   83.9446
theta[Colon(), 22][1] 0.0091 0.0216   0.0005 0.0013  282.8379
theta[Colon(), 22][2] 0.9664 0.0414   0.0009 0.0036  135.4022
theta[Colon(), 22][3] 0.0245 0.0328   0.0007 0.0032  103.1004
theta[Colon(), 23][1] 0.0220 0.0418   0.0009 0.0029  207.5203
theta[Colon(), 23][2] 0.0088 0.0187   0.0004 0.0011  265.6625
theta[Colon(), 23][3] 0.9692 0.0513   0.0011 0.0036  207.4484
theta[Colon(), 24][1] 0.2401 0.0715   0.0016 0.0017 1793.1698
theta[Colon(), 24][2] 0.4293 0.0793   0.0018 0.0026  938.6976
theta[Colon(), 24][3] 0.3306 0.0768   0.0017 0.0031  601.0614
theta[Colon(), 25][1] 0.9232 0.0531   0.0012 0.0056   89.5901
theta[Colon(), 25][2] 0.0102 0.0209   0.0005 0.0012  301.9735
theta[Colon(), 25][3] 0.0666 0.0493   0.0011 0.0055   81.5985

```

**Comment by yebai** [comment]
> I'm a little bit confused by the name change runmodel! -> logp!. I would expect logp! to return the log probability of a particular vi, which shouldn't involve modifying said vi. It looks to me like logp! still returns the vi. Could you expand on this a bit please @yebai ?

I'm not happy with `logp!` either. But a complete solution requires a more careful refactoring, e.g. separate out all places which mutate `vi`. This is planned but not yet implemented.

**Comment by willtebbutt** [comment]
Fair enough. In that case, could we maybe rename `logp!` -> `update_logp!` or something, to make it a bit clearer what's going on? Could we also add a docstring that states precisely what the function does? (Sorry if I've missed it)

File: src/core/VarReplay.jl [hunk 1]
```diff
@@ -103,17 +103,30 @@ mutable struct VarInfo
...
-@generated function Turing.runmodel!(model::Model, vi::VarInfo, spl::AbstractSampler)
+@generated function Turing.logp!(model::Model, vi::VarInfo, spl::AbstractSampler)
     expr_eval_num = spl <: Sampler ?
         :(if :eval_num ∈ keys(spl.info) spl.info[:eval_num] += 1 end) : :()
     return quote
         setlogp!(vi, zero(Real))
         $(expr_eval_num)
-        model(vi, spl)
+        vi.flags["divergent"] = [false]
+        try
```
Comment (xukai92) [inline src/core/VarReplay.jl]: Will this cause any performance issue?

Comment (mohamed82008) [reply src/core/VarReplay.jl]: Yes, this should be optional IMO.

**Comment by yebai** [comment]
Updated, ESS for NUTS (painfully slow at the moment).

```julia
julia> c3= sample(mf, Turing.NUTS(2000, 0.65))
[ Info: [Turing] looking for good initial eps...
[ Info: [Turing] found initial ϵ: 0.24140625
[ Info:  Adapted ϵ = 0.17891128011530977, std = [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]; 1000 iterations is used for adaption.
┌ Warning: Numerical error in gradients. Rejecting current proposal...
└ @ Turing.Core ~/.julia/dev/Turing/src/core/ad.jl:169
┌ Warning: Numerical error in gradients. Rejecting current proposal...
└ @ Turing.Core ~/.julia/dev/Turing/src/core/ad.jl:169
[NUTS] Sampling...100% Time: 2:02:25
[NUTS] Finished with
  Running time        = 7345.092973920997;
  #lf / sample        = 0.0;
  #evals / sample     = 1063.2995;
  pre-cond. metric    = [1.0, 1.0, 1.0, 1.0, 1.0, 1.0,....
Object of type Chains, with data of type 2000×96×1 Array{Union{Missing, Float64},3}

Log evidence      = 0.0
Iterations        = 1:2000
Thinning interval = 1
Chains            = 1
Samples per chain = 2000
internals         = elapsed, epsilon, eval_num, lf_eps, lf_num, lp
parameters        = theta[Colon(), 25][3], theta[Colon(), 4][2], theta[Colon(), 16][2], phi[Colon(), 1][2], theta[Colon(), 14][2], phi[Colon(), 3][2], phi[Colon(), 3][3], phi[Colon(), 2][3], phi[Colon(), 3][1], theta[Colon(), 18][2], theta[Colon(), 22][3], phi[Colon(), 3][5], theta[Colon(), 9][3], theta[Colon(), 24][2], theta[Colon(), 15][3], theta[Colon(), 4][1], theta[Colon(), 8][2], theta[Colon(), 16][3], phi[Colon(), 2][5], theta[Colon(), 21][3], theta[Colon(), 18][1], theta[Colon(), 11][3], theta[Colon(), 2][3], theta[Colon(), 16][1], theta[Colon(), 6][3], theta[Colon(), 24][1], theta[Colon(), 3][3], theta[Colon(), 12][2], theta[Colon(), 7][3], theta[Colon(), 5][2], theta[Colon(), 9][1], theta[Colon(), 7][2], theta[Colon(), 8][1], theta[Colon(), 4][3], theta[Colon(), 15][1], theta[Colon(), 9][2], theta[Colon(), 23][2], theta[Colon(), 15][2], theta[Colon(), 5][3], theta[Colon(), 17][1], theta[Colon(), 7][1], phi[Colon(), 2][2], theta[Colon(), 18][3], theta[Colon(), 24][3], phi[Colon(), 1][5], theta[Colon(), 1][1], theta[Colon(), 20][3], theta[Colon(), 3][1], theta[Colon(), 13][2], theta[Colon(), 1][2], theta[Colon(), 14][1], theta[Colon(), 17][3], theta[Colon(), 20][1], theta[Colon(), 11][1], theta[Colon(), 17][2], theta[Colon(), 11][2], theta[Colon(), 6][1], theta[Colon(), 13][1], theta[Colon(), 22][2], theta[Colon(), 25][1], phi[Colon(), 2][4], theta[Colon(), 6][2], theta[Colon(), 23][3], theta[Colon(), 12][3], theta[Colon(), 2][2], theta[Colon(), 21][1], phi[Colon(), 1][3], theta[Colon(), 19][2], theta[Colon(), 3][2], theta[Colon(), 10][3], phi[Colon(), 2][1], theta[Colon(), 12][1], theta[Colon(), 10][1], theta[Colon(), 10][2], theta[Colon(), 21][2], theta[Colon(), 23][1], theta[Colon(), 14][3], theta[Colon(), 2][1], phi[Colon(), 1][1], phi[Colon(), 1][4], phi[Colon(), 3][4], theta[Colon(), 20][2], theta[Colon(), 13][3], theta[Colon(), 25][2], theta[Colon(), 1][3], theta[Colon(), 5][1], theta[Colon(), 19][3], theta[Colon(), 19][1], theta[Colon(), 8][3], theta[Colon(), 22][1]

parameters
                       Mean    SD   Naive SE  MCSE     ESS   
   phi[Colon(), 1][1] 0.2685 0.0287   0.0006 0.0011  706.0785
   phi[Colon(), 1][2] 0.0693 0.0503   0.0011 0.0030  281.1412
   phi[Colon(), 1][3] 0.0006 0.0012   0.0000 0.0000  738.8822
   phi[Colon(), 1][4] 0.0003 0.0008   0.0000 0.0000  478.2475
   phi[Colon(), 1][5] 0.6613 0.0409   0.0009 0.0022  356.2701
   phi[Colon(), 2][1] 0.0067 0.0172   0.0004 0.0010  283.6553
   phi[Colon(), 2][2] 0.9621 0.0379   0.0008 0.0021  331.7632
   phi[Colon(), 2][3] 0.0191 0.0166   0.0004 0.0019   74.1765
   phi[Colon(), 2][4] 0.0006 0.0015   0.0000 0.0000 1068.9048
   phi[Colon(), 2][5] 0.0115 0.0255   0.0006 0.0017  215.1752
   phi[Colon(), 3][1] 0.0088 0.0513   0.0011 0.0031  267.9855
   phi[Colon(), 3][2] 0.9610 0.0620   0.0014 0.0041  233.4846
   phi[Colon(), 3][3] 0.0171 0.0164   0.0004 0.0020   68.1617
   phi[Colon(), 3][4] 0.0006 0.0015   0.0000 0.0000  942.4492
   phi[Colon(), 3][5] 0.0126 0.0270   0.0006 0.0021  161.9967
 theta[Colon(), 1][1] 0.2496 0.0811   0.0018 0.0036  494.2175
 theta[Colon(), 1][2] 0.3789 0.2711   0.0061 0.0140  376.5666
 theta[Colon(), 1][3] 0.3715 0.2720   0.0061 0.0143  362.8692
 theta[Colon(), 2][1] 0.5788 0.0900   0.0020 0.0027 1084.5385
 theta[Colon(), 2][2] 0.2191 0.1741   0.0039 0.0093  348.7622
 theta[Colon(), 2][3] 0.2021 0.1673   0.0037 0.0085  388.4151
 theta[Colon(), 3][1] 0.2163 0.0637   0.0014 0.0021  929.8895
 theta[Colon(), 3][2] 0.4039 0.3187   0.0071 0.0356   80.1002
 theta[Colon(), 3][3] 0.3798 0.3203   0.0072 0.0358   79.9643
 theta[Colon(), 4][1] 0.2176 0.0735   0.0016 0.0043  288.2934
 theta[Colon(), 4][2] 0.4038 0.3161   0.0071 0.0221  204.2468
 theta[Colon(), 4][3] 0.3786 0.3152   0.0070 0.0217  211.7714
 theta[Colon(), 5][1] 0.0587 0.0464   0.0010 0.0020  563.8994
 theta[Colon(), 5][2] 0.4485 0.3685   0.0082 0.0253  211.5158
 theta[Colon(), 5][3] 0.4928 0.3701   0.0083 0.0250  219.9553
 theta[Colon(), 6][1] 0.0678 0.0571   0.0013 0.0020  850.5224
 theta[Colon(), 6][2] 0.4816 0.3712   0.0083 0.0263  198.7965
 theta[Colon(), 6][3] 0.4506 0.3715   0.0083 0.0254  214.7059
 theta[Colon(), 7][1] 0.9116 0.0628   0.0014 0.0028  516.6733
 theta[Colon(), 7][2] 0.0434 0.0531   0.0012 0.0020  699.0637
 theta[Colon(), 7][3] 0.0451 0.0531   0.0012 0.0024  495.8428
 theta[Colon(), 8][1] 0.0206 0.0324   0.0007 0.0006 2000.0000
 theta[Colon(), 8][2] 0.4865 0.3549   0.0079 0.0189  352.5681
 theta[Colon(), 8][3] 0.4929 0.3555   0.0079 0.0190  349.4413
 theta[Colon(), 9][1] 0.8778 0.0696   0.0016 0.0033  435.8359
 theta[Colon(), 9][2] 0.0635 0.0645   0.0014 0.0030  456.9169
 theta[Colon(), 9][3] 0.0587 0.0636   0.0014 0.0025  660.4352
theta[Colon(), 10][1] 0.6124 0.0889   0.0020 0.0034  701.2793
theta[Colon(), 10][2] 0.1923 0.1588   0.0035 0.0080  398.6490
theta[Colon(), 10][3] 0.1952 0.1594   0.0036 0.0079  406.4469
theta[Colon(), 11][1] 0.0184 0.0317   0.0007 0.0011  806.3091
theta[Colon(), 11][2] 0.5311 0.3670   0.0082 0.0256  206.0394
theta[Colon(), 11][3] 0.4505 0.3669   0.0082 0.0257  204.1278
theta[Colon(), 12][1] 0.0078 0.0328   0.0007 0.0013  600.4953
theta[Colon(), 12][2] 0.5128 0.4202   0.0094 0.0508   68.3534
theta[Colon(), 12][3] 0.4795 0.4210   0.0094 0.0510   68.0130
theta[Colon(), 13][1] 0.5225 0.0891   0.0020 0.0027 1096.0187
theta[Colon(), 13][2] 0.2328 0.1892   0.0042 0.0115  269.7542
theta[Colon(), 13][3] 0.2447 0.1901   0.0043 0.0103  341.9214
theta[Colon(), 14][1] 0.0078 0.0206   0.0005 0.0009  571.3692
theta[Colon(), 14][2] 0.5031 0.3864   0.0086 0.0272  201.0882
theta[Colon(), 14][3] 0.4892 0.3866   0.0086 0.0273  200.5763
theta[Colon(), 15][1] 0.3788 0.0876   0.0020 0.0037  547.1539
theta[Colon(), 15][2] 0.3225 0.2530   0.0057 0.0177  204.1709
theta[Colon(), 15][3] 0.2987 0.2520   0.0056 0.0174  209.8007
theta[Colon(), 16][1] 0.0555 0.0505   0.0011 0.0016  989.8842
theta[Colon(), 16][2] 0.4649 0.3435   0.0077 0.0220  243.5022
theta[Colon(), 16][3] 0.4796 0.3436   0.0077 0.0220  243.1391
theta[Colon(), 17][1] 0.3212 0.0767   0.0017 0.0025  923.8209
theta[Colon(), 17][2] 0.3396 0.2729   0.0061 0.0177  237.8483
theta[Colon(), 17][3] 0.3392 0.2737   0.0061 0.0180  231.0518
theta[Colon(), 18][1] 0.9504 0.0448   0.0010 0.0020  520.6452
theta[Colon(), 18][2] 0.0247 0.0335   0.0007 0.0015  488.9008
theta[Colon(), 18][3] 0.0249 0.0339   0.0008 0.0012  796.5029
theta[Colon(), 19][1] 0.6986 0.0846   0.0019 0.0032  705.5060
theta[Colon(), 19][2] 0.1551 0.1265   0.0028 0.0062  413.7902
theta[Colon(), 19][3] 0.1462 0.1287   0.0029 0.0055  551.0321
theta[Colon(), 20][1] 0.7344 0.0882   0.0020 0.0036  616.1414
theta[Colon(), 20][2] 0.1337 0.1187   0.0027 0.0060  395.0118
theta[Colon(), 20][3] 0.1319 0.1175   0.0026 0.0037 1029.0750
theta[Colon(), 21][1] 0.7951 0.0926   0.0021 0.0043  457.9824
theta[Colon(), 21][2] 0.0990 0.0983   0.0022 0.0048  428.1273
theta[Colon(), 21][3] 0.1058 0.1004   0.0022 0.0025 1639.8742
theta[Colon(), 22][1] 0.5864 0.0827   0.0018 0.0022 1375.2300
theta[Colon(), 22][2] 0.2168 0.1634   0.0037 0.0105  241.3385
theta[Colon(), 22][3] 0.1968 0.1606   0.0036 0.0106  231.5896
theta[Colon(), 23][1] 0.1422 0.0644   0.0014 0.0025  650.0146
theta[Colon(), 23][2] 0.4300 0.3121   0.0070 0.0142  481.0965
theta[Colon(), 23][3] 0.4278 0.3157   0.0071 0.0152  433.7158
theta[Colon(), 24][1] 0.0092 0.0353   0.0008 0.0015  573.7085
theta[Colon(), 24][2] 0.4610 0.3982   0.0089 0.0254  245.4244
theta[Colon(), 24][3] 0.5298 0.3981   0.0089 0.0252  249.4068
theta[Colon(), 25][1] 0.9243 0.0564   0.0013 0.0029  380.9910
theta[Colon(), 25][2] 0.0382 0.0466   0.0010 0.0019  622.2045
theta[Colon(), 25][3] 0.0375 0.0460   0.0010 0.0023  399.8860


```

**Comment by yebai** [comment]
Places that mutate `v::VarIngo` in `logp!`:

- https://github.com/TuringLang/Turing.jl/blob/c5ca896e27d94353a3466317e1372994ca3fda48/src/inference/pgibbs.jl#L155
- https://github.com/TuringLang/Turing.jl/blob/c5ca896e27d94353a3466317e1372994ca3fda48/src/inference/pgibbs.jl#L159
- https://github.com/TuringLang/Turing.jl/blob/c5ca896e27d94353a3466317e1372994ca3fda48/src/inference/pgibbs.jl#L172
- https://github.com/TuringLang/Turing.jl/blob/c5ca896e27d94353a3466317e1372994ca3fda48/src/inference/hmc.jl#L270

**Comment by yebai** [comment]
I think it's better resolve this issue later, e.g. after https://github.com/TuringLang/Turing.jl/issues/634 and https://github.com/TuringLang/Turing.jl/issues/723.

**Comment by xukai92** [comment]
What's the plan for this issue.

**Comment by yebai** [comment]
This is probably obsolete now given the recent improvements of AHMC - some numerical issues are already resolved by more robust adaption I think.

**Pull request closed without merging**